File: packages/engine/src/pricing.ts
import { readFileSync } from "node:fs";
import type { UsageRecord } from "./types";
import bundledPricingTable from "./pricing.json";

/**
 * Cennik — spec.md sekcja 3.3.
 * Ceny w USD za 1M tokenów, per model.
 */
export interface ModelPricing {
  inputPer1M: number;
  outputPer1M: number;
  cacheReadPer1M: number;
  cacheWritePer1M: number;
}

export type PricingTable = Record<string, ModelPricing>;

// Ceny bazowe (input/output) wg oficjalnego cennika Anthropic (stan: wrzesień 2026).
// Ceny cache write/read wyliczone wg udokumentowanej reguły Anthropic dla cache
// o TTL 5 minut: cache write = 1.25x cena input, cache read = 0.1x cena input.
const defaultPricingTable = bundledPricingTable as PricingTable;

// Claude Code zapisuje ten pseudo-model dla komunikatów generowanych lokalnie
// (bez wywołania API) — zerowy koszt jest tu poprawny, nie "nieznany".
const SYNTHETIC_MODEL = "<synthetic>";

// Claude Code zapisuje samą nazwę rodziny, gdy użytkownik wybrał model aliasem
// (`/model sonnet`). Alias wskazuje aktualnie najnowszy model danej rodziny.
const MODEL_FAMILY_ALIASES: Record<string, string> = {
  sonnet: "claude-sonnet-5",
  opus: "claude-opus-5",
  haiku: "claude-haiku-4-5",
  fable: "claude-fable-5-1",
};

/**
 * Sprowadza ID modelu z logów do klucza tabeli cen: zdejmuje prefix dostawcy
 * `anthropic/` i sufiks okna kontekstu `[1m]` (oba dodawane przez bramki
 * API / `/model anthropic/...[1m]`), rozwiązuje aliasy rodzin. ID innych
 * dostawców (np. `openai/...`) zwraca bez zmian.
 */
export function normalizeModelId(model: string): string {
  let id = model;
  if (id.startsWith("anthropic/")) {
    id = id.slice("anthropic/".length);
  }
  if (id.endsWith("[1m]")) {
    id = id.slice(0, -"[1m]".length);
  }
  return MODEL_FAMILY_ALIASES[id] ?? id;
}

/**
 * Wylicza `costUsd` dla rekordu na podstawie tabeli cen i zwraca nową kopię
 * rekordu z ustawionym polem. Nigdy nie rzuca wyjątku: dla nieznanego modelu
 * zwraca `costUsd = 0` i (opcjonalnie) informuje przez `onUnknownModel`
 * (spec.md 3.3 — "nigdy nie przerywa działania").
 */
export function applyPricing(
  record: UsageRecord,
  table: PricingTable,
  onUnknownModel?: (model: string) => void,
): UsageRecord {
  if (record.model === SYNTHETIC_MODEL) {
    return { ...record, costUsd: 0 };
  }

  const pricing = table[record.model] ?? table[normalizeModelId(record.model)];
  if (!pricing) {
    onUnknownModel?.(record.model);
    return { ...record, costUsd: 0 };
  }

  const costUsd =
    (record.tokensInput / 1_000_000) * pricing.inputPer1M +
    (record.tokensOutput / 1_000_000) * pricing.outputPer1M +
    (record.tokensCacheRead / 1_000_000) * pricing.cacheReadPer1M +
    (record.tokensCacheWrite / 1_000_000) * pricing.cacheWritePer1M;

  return { ...record, costUsd };
}

/**
 * Wczytuje tabelę cen: domyślną wbudowaną, scaloną z opcjonalnym plikiem
 * użytkownika (nadpisanie per-model, patrz spec.md 3.3). Jeśli `customPath`
 * nie istnieje lub zawiera niepoprawny JSON, cicho pomija nadpisanie i zwraca
 * samą domyślną tabelę — nigdy nie rzuca wyjątku.
 */
export function loadPricingTable(customPath?: string): PricingTable {
  if (!customPath) {
    return { ...defaultPricingTable };
  }

  let custom: PricingTable;
  try {
    custom = JSON.parse(readFileSync(customPath, "utf-8")) as PricingTable;
  } catch {
    return { ...defaultPricingTable };
  }

  return { ...defaultPricingTable, ...custom };
}

File: packages/intellij-plugin/engine/src/main/kotlin/dev/bredson/tokentracker/engine/Pricing.kt
package dev.bredson.tokentracker.engine

import java.nio.file.Path
import kotlin.io.path.readText
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/** Ceny w USD za 1M tokenów — ten sam format co `pricing.json` silnika TS. */
@Serializable
data class ModelPricing(
    val inputPer1M: Double,
    val outputPer1M: Double,
    val cacheReadPer1M: Double,
    val cacheWritePer1M: Double,
)

typealias PricingTable = Map<String, ModelPricing>

private val json = Json { ignoreUnknownKeys = true }

private const val SYNTHETIC_MODEL = "<synthetic>"

private val MODEL_FAMILY_ALIASES = mapOf(
    "sonnet" to "claude-sonnet-5",
    "opus" to "claude-opus-5",
    "haiku" to "claude-haiku-4-5",
    "fable" to "claude-fable-5-1",
)

private val bundledTable: PricingTable by lazy {
    val text = checkNotNull(ModelPricing::class.java.getResource("/pricing.json")) {
        "pricing.json missing from engine resources"
    }.readText()
    json.decodeFromString<Map<String, ModelPricing>>(text)
}

/** Zdejmuje `anthropic/` i `[1m]`, rozwiązuje aliasy rodzin; inne ID zwraca bez zmian. */
fun normalizeModelId(model: String): String {
    val id = model.removePrefix("anthropic/").removeSuffix("[1m]")
    return MODEL_FAMILY_ALIASES[id] ?: id
}

/** Wbudowana tabela scalona z opcjonalnym plikiem użytkownika; nigdy nie rzuca. */
fun loadPricingTable(customPath: Path? = null): PricingTable {
    if (customPath == null) return bundledTable
    val custom = try {
        json.decodeFromString<Map<String, ModelPricing>>(customPath.readText())
    } catch (_: Exception) {
        return bundledTable
    }
    return bundledTable + custom
}

/** Nowa kopia rekordu z `costUsd`; nieznany model → 0 i wywołanie `onUnknownModel`. */
fun applyPricing(
    record: UsageRecord,
    table: PricingTable,
    onUnknownModel: ((String) -> Unit)? = null,
): UsageRecord {
    if (record.model == SYNTHETIC_MODEL) return record.copy(costUsd = 0.0)

    val pricing = table[record.model] ?: table[normalizeModelId(record.model)]
    if (pricing == null) {
        onUnknownModel?.invoke(record.model)
        return record.copy(costUsd = 0.0)
    }

    val cost = record.tokensInput / 1_000_000.0 * pricing.inputPer1M +
        record.tokensOutput / 1_000_000.0 * pricing.outputPer1M +
        record.tokensCacheRead / 1_000_000.0 * pricing.cacheReadPer1M +
        record.tokensCacheWrite / 1_000_000.0 * pricing.cacheWritePer1M
    return record.copy(costUsd = cost)
}
